docker-metadata.lua

```diff
@@ -0,0 +1,113 @@
+local M = {}
+
+-- Directory where Docker stores container data
+M.DOCKER_VAR_DIR = '/var/lib/docker/containers/'
+
+-- Docker container configuration file
+M.DOCKER_CONTAINER_CONFIG_FILE = '/config.v2.json'
+
+-- Cache Time-To-Live in seconds
+M.CACHE_TTL_SEC = 300
+
+-- Table defining patterns to extract metadata from Docker config file
+M.DOCKER_CONTAINER_METADATA = {
+  ['docker_container_name'] = '\"Name\":\"/?(.-)\"',  -- Extract container name
+  ['docker_container_image'] = '\"Image\":\"/?(.-)\"',  -- Extract container image name
+  ['docker_container_started'] = '\"StartedAt\":\"/?(.-)\"'  -- Extract container start time
+}
+
+-- Cache to store metadata for containers
+M.cache = {}
+
+local debug_mode = os.getenv("DEBUG_MODE") == "true"
+
+-- Function to print debug messages if debug mode is enabled
+local function debug_print(...)
+  if debug_mode then
+    print(...)
+  end
+end
+
+-- Function to extract container ID from log tag
+function M.get_container_id_from_tag(tag)
+  debug_print("Getting container ID from tag:", tag)
+  local container_id = tag:match('containers%.([a-f0-9]+)')
+  debug_print("Container ID:", container_id)
+  return container_id
+end
+
+-- Function to read and extract metadata from Docker config file
+function M.get_container_metadata_from_disk(container_id)
+  local docker_config_file = M.DOCKER_VAR_DIR .. container_id .. M.DOCKER_CONTAINER_CONFIG_FILE
+  debug_print("Reading metadata from:", docker_config_file)
+
+  local fl = io.open(docker_config_file, 'r')
+  if fl == nil then
+    debug_print("Failed to open file:", docker_config_file)
+    return nil
+  end
+
+  local data = { time = os.time() }
+
+  for line in fl:lines() do
+    for key, regex in pairs(M.DOCKER_CONTAINER_METADATA) do
+      local match = line:match(regex)
+      if match then
+        data[key] = match
+        debug_print("Found metadata:", key, match)
+      end
+    end
+  end
+  fl:close()
+
+  if next(data) == nil then
+    debug_print("No metadata found in file:", docker_config_file)
+    return nil
+  else
+    debug_print("Metadata extracted for container:", container_id)
+    return data
+  end
+end
+
+-- Function to enrich log records with Docker metadata
+function M.enrich_with_docker_metadata(tag, timestamp, record)
+  debug_print("Enriching record with tag:", tag)
+
+  local container_id = M.get_container_id_from_tag(tag)
+  if not container_id then
+    debug_print("No container ID found for tag:", tag)
+    return 0, 0, 0
+  end
+
+  local new_record = record
+  new_record['docker_container_id'] = container_id
+  new_record['message'] = record.log
```

Comment:
why do we change the record in lua script instead of using `modify` filter in the fluentbit?
https://docs.fluentbit.io/manual/pipeline/filters/modify

Reply:
We use a Lua script instead of the Fluent Bit modify filter for complex data enrichment that modify can’t handle. The Lua script reads Docker config files for container metadata (e.g., container name, image, start time), which the modify filter can't access.

Reply:
I see, cool! 💪🏼 

I was referring in that commen to the `new_record['message'] = record.log` part. It can be managed by the `modify` filter, and using it would make it easier for users to change it if they want a custom config."

Reply:
You right. The Rename log message operation is now included at the start of the [FILTER] block instead of new_record['message'] = record.log part here.